lib/gemstash/compact_index_builder.rb

```diff
@@ -0,0 +1,257 @@
+                row[:platform],
+                nil, # sha256
+                row[:info_checksum],
+                nil, # dependencies
+                nil, # version.required_ruby_version,
+                nil, # version.required_rubygems_version
+              )
+            end
+          )
+        end
+      end
+
+      def key
+        :versions
+      end
+    end
+
+    # Comment
+    class Info < CompactIndexBuilder
+      def initialize(auth, name)
+        super(auth)
+        @name = name
+      end
+
+      def fetch_resource
+        storage.resource("info/#{@name}")
+      end
+
+      def build_result
+        @result = CompactIndex.info(requirements_and_dependencies)
+      end
+
+    private
+
+      def requirements_and_dependencies
+        group_by_columns = "number, platform, sha256, info_checksum, required_ruby_version, required_rubygems_version, versions.created_at"
+
+        dep_req_agg = "string_agg(dependencies.requirements, '@' ORDER BY dependencies.rubygem_name, dependencies.id) as dep_req_agg"
+
+        dep_name_agg = "string_agg(dependencies.rubygem_name, ',' ORDER BY dependencies.rubygem_name) AS dep_name_agg"
+
+        DB::Rubygem.db[<<~SQL.squish, @name].
+          SELECT #{group_by_columns}, #{dep_req_agg}, #{dep_name_agg}
+          FROM rubygems
+            LEFT JOIN versions ON versions.rubygem_id = rubygems.id
+            LEFT JOIN dependencies ON dependencies.version_id = versions.id
+          WHERE rubygems.name = ? AND versions.indexed = true
+          GROUP BY #{group_by_columns}
+          ORDER BY versions.created_at, number, platform, dep_name_agg
+        SQL
+          map do |row|
+          reqs = row[:dep_req_agg]&.split("@")
+          dep_names = row[:dep_name_agg]&.split(",")
+
+          raise "Dependencies and requirements are not the same size:\n  reqs: #{reqs.inspect}\n  dep_names: #{dep_names.inspect}\n  row: #{row.inspect}" if dep_names&.size != reqs&.size
+
+          deps = []
+          if reqs
+            dep_names.zip(reqs).each do |name, req|
+              deps << CompactIndex::Dependency.new(name, req)
+            end
+          end
+
+          CompactIndex::GemVersion.new(
+            row[:number],
+            row[:platform],
+            row[:sha256],
+            nil, # info_checksum
+            deps,
+            row[:required_ruby_version],
+            row[:required_rubygems_version]
+          )
+        end
+      end
+
+      def key
+        :info
+      end
+    end
+
+    # Comment
+    class Names < CompactIndexBuilder
+      def fetch_resource
+        storage.resource("names")
+      end
+
+      def build_result
+        names = DB::Rubygem.db[<<~SQL.squish].map {|row| row[:name] }
+          SELECT name
+          FROM rubygems
+          LEFT JOIN versions ON versions.rubygem_id = rubygems.id
+          WHERE versions.indexed = true
+          GROUP BY name
+          HAVING COUNT(versions.id) > 0
+          ORDER BY name
```

Comment:
I wonder, can this become an inner join rather than a LEFT JOIN? That is, excluding rubygems rows that have no related row in versions. Then, we'd not need the `HAVING COUNT(versions.id) > 0`.

Am I missing something here?